subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/AbstractFileWatcherRegistryFactory.java

```diff
@@ -46,7 +50,25 @@ public FileWatcherRegistry createFileWatcherRegistry(FileWatcherRegistry.ChangeH
         }
     }
 
+    @Override
+    public boolean isAvailable() {
+        try {
+            init();
+            return true;
+        } catch (NativeIntegrationUnavailableException e) {
+            LOGGER.info("Native file-system watching is not available for the current operating system", e);
+            return false;
+        }
+    }
+
     protected abstract FileWatcher createFileWatcher(BlockingQueue<FileWatchEvent> fileEvents) throws InterruptedException;
 
     protected abstract FileWatcherUpdater createFileWatcherUpdater(FileWatcher watcher);
+
+    /**
+     * Initializes the native parts of file-system watching.
+     *
+     * @throws NativeIntegrationUnavailableException if the native parts of file-system watching are not availabe on the current operating system. E.g. some older Linux variants are not supported.
+     */
+    protected abstract void init() throws NativeIntegrationUnavailableException;
```

Comment:
I feel that having this side-effect-only method is a code smell. We now call `Native.get(*FileEventFunctions.class)` twice in each implementation. I'd rather have a `getFileEventFunctions()` method that returns the `AbstractFileEventFunctions` instance (the individual implementations could use more specific return types).

Reply:
I moved the initialization to the constructor.